test/unit/lib/utils/fs/safeMoveFile.test.js

```diff
@@ -0,0 +1,160 @@
+
+describe('#safeMoveFile()', () => {
+  let renameStub;
+  let copyFileStub;
+  let unlinkStub;
+
+  beforeEach(async () => {
+    renameStub = sinon.stub(fsp, 'rename');
+    copyFileStub = sinon.stub(fsp, 'copyFile');
+    unlinkStub = sinon.stub(fsp, 'unlink');
+    // Allow the fs calls to act as normal until we want to force the rename to fail
+    renameStub.callThrough();
+    copyFileStub.callThrough();
+    unlinkStub.callThrough();
+
+    // Ensure the cache directory exists so that the rename doesn't fail
+    fse.ensureDirSync(sourcePath);
+    fse.ensureDirSync(destinationPath);
+
+    // Write a test file to rename
+    await fse.writeFile(sourceFile, 'source data');
+  });
+
+  afterEach(() => {
+    // Clean up test directories after each test so they don't interfere with each other
+    fse.removeSync(destinationPath);
+    fse.removeSync(sourcePath);
+    // Reset stubbed methods
+    fsp.rename.restore();
+    fsp.copyFile.restore();
+    fsp.unlink.restore();
+  });
+
+  /**
+   * Run the test suite with the provided post assertion function.
+   * The post assertion function will be called after each test scenario is executed
+   * so that the test scenarios can be reused.
+   *
+   * @param {*} postAssertion the function that is called after every test scenario
+   */
+  const runTestScenariosWithPostAssertion = (postAssertion) => {
+    describe('when the file does not exist by the time the zip is generated', () => {
+      it('should write a file', async () => {
+        await safeMoveFile(sourceFile, destinationFile);
+
+        const exists = fse.existsSync(destinationFile);
+        expect(exists).to.be.true;
+        postAssertion();
+      });
+    });
+
+    describe('when the file already exists by the time the zip is generated', () => {
+      beforeEach(async () => {
+        // Create an existing file that is at the destination
+        fse.ensureFileSync(destinationFile);
+        fse.writeFileSync(destinationFile, 'existing destination data');
+      });
+
+      describe('when the cached file is not open', () => {
+        it('should overwrite the cached file with a valid zip file', async () => {
+          await safeMoveFile(sourceFile, destinationFile);
+
+          // Check that the file was actually overwritten
+          const cachedData = fse.readFileSync(destinationFile).toString();
+          expect(cachedData).not.to.eq('existing destination data');
+
+          postAssertion();
+        });
+      });
+    });
+  };
+
+  describe('when the temporary directory is not on a different device to the cache directory', () => {
+    runTestScenariosWithPostAssertion(() => {
+      // Happy path: Only rename is called
+      expect(renameStub).to.have.be.calledOnce;
+      expect(copyFileStub).to.not.have.been.called;
+      expect(unlinkStub).to.not.have.been.called;
+    });
+  });
+
+  describe('when the temporary directory is on a different device to the cache directory', () => {
```

Comment:
Let's rename to `when file is moved to different device`

Reply:
Done!